src/wagtail_bynder/utils.py

```diff
@@ -2,47 +2,137 @@
 import os
 
 from io import BytesIO
+from tempfile import NamedTemporaryFile
 
 import requests
 
 from asgiref.local import Local
 from bynder_sdk import BynderClient
 from django.conf import settings
-from django.core.files.uploadedfile import InMemoryUploadedFile
+from django.core.files.uploadedfile import InMemoryUploadedFile, UploadedFile
+from django.template.defaultfilters import filesizeformat
+from PIL import Image
+from wagtail.images.models import WagtailImageFieldFile
 from wagtail.models import Collection
 
-
-_DEFAULT_COLLECTION = Local()
-
-
-class DownloadedFile(BytesIO):
-    name: str
-    size: int
-    content_type: str | None
-    charset: str | None
+from .exceptions import BynderAssetFileTooLarge
 
 
-def download_file(url: str) -> DownloadedFile:
-    raw_bytes = requests.get(url, timeout=20).content
-    f = DownloadedFile(raw_bytes)
-    f.name = os.path.basename(url)
-    f.size = len(raw_bytes)
-    f.content_type, f.charset = mimetypes.guess_type(f.name)
-    return f
+_DEFAULT_COLLECTION = Local()
 
 
 def download_asset(url: str) -> InMemoryUploadedFile:
-    f = download_file(url)
+    name = os.path.basename(url)
+    max_filesize = int(getattr(settings, "BYNDER_MAX_DOWNLOAD_FILE_SIZE", 5242880))
+
+    # Stream file to
+    file = BytesIO()
+    for line in requests.get(url, timeout=20, stream=True):
+        file.write(line)
+        if file.tell() > max_filesize:
+            file.truncate()
+            raise BynderAssetFileTooLarge(
+                f"File '{name}' exceeded BYNDER_MAX_DOWNLOAD_FILE_SIZE, which is set to {filesizeformat(max_filesize)}."
+            )
+
+    size = file.tell()
+    file.seek(0)
+
+    content_type, charset = mimetypes.guess_type(name)
     return InMemoryUploadedFile(
-        f,
-        name=f.name,
+        file,
         field_name="file",
-        size=f.size,
-        charset=f.charset,
-        content_type=f.content_type,
+        name=name,
+        content_type=content_type,
+        size=size,
+        charset=charset,
     )
 
 
+def reduce_image_size(
+    file: UploadedFile | WagtailImageFieldFile,
+    max_width: int | None = None,
+    max_height: int | None = None,
+):
+    default_max_width = int(getattr(settings, "BYNDER_IMAGE_MAX_WIDTH", 3500))
+    default_max_height = int(getattr(settings, "BYNDER_IMAGE_MAX_HEIGHT", 3500))
+    _max_width = default_max_width if max_width is None else max_width
+    _max_height = default_max_height if max_height is None else max_height
+
+    name, ext = os.path.splitext(file.name)
+    resized = None
+    new_content_type = None
+    new_filename = None
+
+    with Image.open(file) as image:
```

Comment:
Want to swap Willow for Pillow here, and also utilize Wagtail's `WAGTAILIMAGES_FORMAT_CONVERSIONS` setting to decide which formats to convert to